### ISSUE TITLE
Functions to open a temporary file (temporary directory)

### ISSUE THREAD

**Issue by certik** [fortran-lang/stdlib#126]
Similar to Python's [mkstemp](https://docs.python.org/3.8/library/tempfile.html#tempfile.mkstemp) and [mkdtemp](https://docs.python.org/3.8/library/tempfile.html#tempfile.mkdtemp).

**Comment by epagone** [comment]
I might be missing something obvious, but why can't we just use `open(status='SCRATCH',...)`?

**Comment by certik** [comment]
We could. I haven't thought of that as an option.

**Comment by urbanjost** [comment]
There are some similiar routines described in
[M_io](https://urbanjost.github.io/general-purpose-fortran/download/tmp/html/BOOK_M_io.html) and M_io.f90 and M_path.f90 in [GPF](https://urbanjost.github.com/general-purpose-fortran). There are seperate routines for a true status='scratch' file and a unique scratchfile name; and a discussion of the pros and cons of how status='scratch' is implemented (there is a big variation in implementations) at [wiki](http://fortranwiki.org/fortran/show/Scratch+Files) that brings up a few relevant points about how depending on status='scratch'  has pitfalls because of bad implementations.   I really like the idea of making a scratch directory, especially if it was cleaned up automatically like a scratch file should be;  I do not have anything like that. The python implementation does not clean up at termination automatically.

**Comment by certik** [comment]
Python's `mkdtemp` does not clean up, but [TemporaryDirectory](https://docs.python.org/3.8/library/tempfile.html#tempfile.TemporaryDirectory) does.

**Comment by jvdp1** [comment]
> I really like the idea of making a scratch directory, especially if it was cleaned up automatically like a scratch file should be; I do not have anything like that.

@urbanjost: I like this idea too. Would it be possible with derived types and a `final` procedure?

**Comment by certik** [comment]
Yes, we should use a derived type with a finalizer for this. I updated the issue description with links to temporary files and directories implementation in Python, Rust and Julia.
